batch/batch/worker.py

```diff
@@ -81,30 +84,94 @@ def free(self, port):
         self.ports.put_nowait(port)
 
 
-async def docker_call_retry(f, *args, **kwargs):
-    delay = 0.1
+def docker_call_retry(timeout, name):
+    async def _wrapper(f, *args, **kwargs):
+        delay = 0.1
+        local_timeout = timeout
+        while True:
+            try:
+                return await asyncio.wait_for(f(*args, **kwargs), local_timeout)
+            except DockerError as e:
+                # 408 request timeout, 503 service unavailable
+                if e.status == 408 or e.status == 503:
+                    log.exception(f'in docker call to {f.__name__} for {name}, retrying', stack_info=True)
+                # DockerError(500, 'Get https://registry-1.docker.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
+                # DockerError(500, 'error creating overlay mount to /var/lib/docker/overlay2/545a1337742e0292d9ed197b06fe900146c85ab06e468843cd0461c3f34df50d/merged: device or resource busy'
+                elif e.status == 500 and ("request canceled while waiting for connection" in e.message or
+                                          re.match("error creating overlay mount.*device or resource busy", e.message)):
+                    log.exception(f'in docker call to {f.__name__} for {name}, retrying', stack_info=True)
+                else:
+                    raise
+            except aiohttp.client_exceptions.ServerDisconnectedError:
+                log.exception(f'in docker call to {f.__name__}, retrying', stack_info=True)
+            except asyncio.TimeoutError:
+                local_timeout *= 2
```

Comment:
What's the justification for this?

Reply:
I thought it would be good to increase the timeout in case we're trying to pull a really large image or something like that where it just legitimately is going to take a long time to complete the operation and not transitively stuck.